app/assets/javascripts/discourse/app/models/user.js

```diff
@@ -1133,6 +1139,68 @@ User.reopenClass(Singleton, {
   },
 });
 
+// user status tracking
+User.reopen(Evented, {
+  _clearStatusTimerId: null,
+
+  // always call stopTrackingStatus() when done with a user
+  trackStatus() {
+    this.addObserver("status", this, "_statusChanged");
+
+    if (this.isCurrent) {
+      this.appEvents.on("current-user-status:changed", (status) => {
+        this.set("status", status);
+      });
+    }
+
+    if (this.status && this.status.ends_at) {
+      this._scheduleStatusClearing(this.status.ends_at);
+    }
+  },
+
+  stopTrackingStatus() {
+    this.removeObserver("status", this, "_statusChanged");
+    if (this.isCurrent) {
+      this.appEvents.off("current-user-status:changed");
+    }
+    this._unscheduleStatusClearing();
+  },
+
+  _statusChanged(sender, key) {
+    this.trigger("status-changed");
+
+    const status = this.get(key);
+    if (status && status.ends_at) {
+      this._scheduleStatusClearing(status.ends_at);
+    } else {
+      this._unscheduleStatusClearing();
+    }
+  },
+
+  _scheduleStatusClearing(endsAt) {
+    if (this._clearStatusTimerId) {
+      this._unscheduleStatusClearing();
+    }
+
+    const utcNow = moment.utc();
+    const remaining = moment.utc(endsAt).diff(utcNow, "milliseconds");
+
+    this._clearStatusTimerId = setTimeout(
```

Comment:
we don't want setTimeout in Ember, it doesn’t play well with runloop AFAIK, should be replaced by later most likely:

https://api.emberjs.com/ember/release/functions/@ember%2Frunloop/later